cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py

```diff
@@ -102,12 +101,28 @@ def check_for_upgrade_needed():
     add_rbac_roles()
     set_state('reconfigure.authentication.setup')
     remove_state('authentication.setup')
+    if snap_resources_changed():
+        set_upgrade_needed()
+
+
+def snap_resources_changed():
+    '''
+    Check if the snapped resources have changed. The first time this method is
+    called will report no change.
+
+    Returns: True in case a snap resource file has changed
 
+    '''
+    db = unitdata.kv()
     resources = ['kubectl', 'kube-apiserver', 'kube-controller-manager',
                  'kube-scheduler', 'cdk-addons']
     paths = [hookenv.resource_get(resource) for resource in resources]
-    if any_file_changed(paths):
-        set_upgrade_needed()
+    if db.get('snap.resources.fingerprint.initialised'):
+        return any_file_changed(paths)
+    else:
+        db.set('snap.resources.fingerprint.initialised', True)
+        any_file_changed(paths)
+        return False
```

Comment:
Ah shoot. There's a corner case I'm worried about:
```
juju upgrade-charm kubernetes-master --resource kube-apiserver=my-kube-apiserver.snap
```
If both the charm upgrade and the new resource are processed in a single `upgrade-charm` hook, then this will return False, no upgrade will occur, and the admin won't be informed that their new resource isn't running.

I don't think we can safely return False here without proving that the above scenario can't happen.

Could we instead set both `kubernetes-master.upgrade-needed` and `kubernetes-master.upgrade-specified` in this case? I think I like the behavior of that better:

1. If no new resources were attached, we don't annoy the admin with an unnecessary action.
2. If new resources were attached, we don't leave the cluster sneakily running the old one.

Dunno, what do you think?

Reply:
That is a valid point. I did not think of this upgrade scenario.

We are getting there!